nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/TrashedStateMigrator.java

```diff
@@ -110,19 +111,27 @@ protected void migrateSession(CoreSession session) {
         checkShutdownRequested();
 
         // compute all deleted doc id refs
-        List<IdRef> deletedRefs = deletedMaps.stream() //
-                                             .map(map -> (String) map.get(ECM_UUID))
-                                             .map(IdRef::new)
-                                             .collect(Collectors.toList());
+        List<String> ids = deletedMaps.stream().map(map -> (String) map.get(ECM_UUID)).collect(toList());
 
         checkShutdownRequested();
 
         // set ecm:isTrashed property by batch
-        int size = deletedRefs.size();
+        int size = ids.size();
         int i = 0;
         reportProgress(session.getRepositoryName(), "Setting isTrashed property", i, size);
-        for (IdRef deletedRef : deletedRefs) {
-            session.setDocumentSystemProp(deletedRef, SYSPROP_IS_TRASHED, Boolean.TRUE);
+        LifeCycleService lifeCycleService = Framework.getService(LifeCycleService.class);
+        for (String id : ids) {
+            Document doc = ((AbstractSession) session).getSession().getDocumentByUUID(id);
```

Comment:
What are you using the low-level `Session` here? Worth a comment.

Reply:
I need to have `Document` in order to follow the transition without hitting the backward compat mechanism, I will add a comment.

Reply:
I would have preferred using the normal `CoreSession` and adding a new flag in `followTransition`'s `option` map to bypass the backward compat mechanism.
Using the low-level `Session` should be strictly for internal use of the `CoreSession` and low-level services only. This is a violation of layering.